### PR TITLE
chore(ci): Upload build artifacts on releases

### PR DESCRIPTION
**Pull request by kevaundray**
# Description

This currently uploads the static libraries (WASM, linux, Mac) , the dest folder which contains the transpiled javascript and a barretenberg.wasm file.

What we want long term is:

- No static libraries (instead we have binaries)
- No dest folder, we upload to NPM (this is pending an NPM token and a modification to change the npm version when releases are made)
- Barretenberg.wasm (We still want to keep the barretenberg.wasm binary being upload)

Please provide a paragraph or two giving a summary of the change, including relevant motivation and context.

# Checklist:

- [ ] I have reviewed my diff in github, line by line.
- [ ] Every change is related to the PR description.
- [ ] The branch has been merged with/rebased against the head of its merge target.
- [ ] There are no unexpected formatting changes, superfluous debug logs, or commented-out code.
- [ ] There are no circuit changes, OR a cryptographer has been assigned for review.
- [ ] New functions, classes, etc. have been documented according to the doxygen comment format. Classes and structs must have `@brief` describing the intended functionality.
- [ ] If existing code has been modified, such documentation has been added or updated.
- [ ] No superfluous `include` directives have been added.
- [ ] I have [linked](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue) to any issue(s) it resolves.
- [ ] I'm happy for the PR to be merged at the reviewer's next convenience.

### CODE REVIEW

**Comment by kevaundray** [comment]
The workflows are dependent, so we do not upload to github unless both MacOS and Linux build. We also don't build unless a release was successfully made. See the `needs` field in the yaml file.

The linux runner builds for both linux and WASM which is why we require WASI for Linux and not for Mac. Since we don't care too much about when the artifacts are uploaded right now, I'm using the free github version, they take about 30-40 minutes to build and publish to GitHub.

A c5ad.16xlarge AWS machine can put this down to 1 minute of build time, but it also costs about $2.7 an hour. Note that we can also take advantage of github caching which would mean that build times may only be bad if you modify code in field arithmetic for example (From conversations, I think its actually not this optimistic)

File: .github/workflows/release-please.yml [hunk 1]
```diff
@@ -26,3 +29,100 @@ jobs:
...
+      - name: Upload artifacts
+        uses: actions/upload-artifact@v2
+        with:
+          name: release-mac
+          path: macos-static-lib.tar
+
+  release:
+    name: Publish
+    needs: [build-linux, build-mac, release-please]
+    runs-on: ubuntu-large2
```
Comment (Rumata888) [inline .github/workflows/release-please.yml]: Why do we need a large ubuntu machine for downloading and uploading artifacts?

Comment (kevaundray) [inline .github/workflows/release-please.yml]:
```suggestion
    runs-on: ubuntu-latest
```

File: .github/workflows/release-please.yml [hunk 2]
```diff
@@ -26,3 +29,100 @@ jobs:
             VERSION
             barretenberg.nix
             barretenberg-wasm.nix
+
+  build-linux:
+    name: Build on Ubuntu
+    needs: [release-please]
+    runs-on: ubuntu-latest
```
Comment (charlielye) [inline .github/workflows/release-please.yml:27]: Can we build on ubuntu 20 LTS (focal) to ensure compatability with mainframe.

File: .github/workflows/release-please.yml [hunk 3]
```diff
@@ -26,3 +29,100 @@ jobs:
...
+    steps:
+      - name: Checkout
+        uses: actions/checkout@v3
+      - name: Install Bleeding edge CMake
```
Comment (charlielye) [inline .github/workflows/release-please.yml]: Is stock cmake not sufficient? Do you know what features we use that aren't in stock install?

File: .github/workflows/release-please.yml [hunk 4]
```diff
@@ -26,3 +29,100 @@ jobs:
...
+        run: |
+              sudo apt -y remove --purge cmake
+              sudo snap install cmake --classic
+      - name: Install Yarn
+        run: |
+          curl -sS https://dl.yarnpkg.com/debian/pubkey.gpg | sudo apt-key add -
+          echo "deb https://dl.yarnpkg.com/debian/ stable main" | sudo tee /etc/apt/sources.list.d/yarn.list
+          sudo apt -y update && sudo apt -y install yarn
+      - name: Install Clang/Ninja/Binaryen
```
Comment (charlielye) [inline .github/workflows/release-please.yml]: We don't need binaryen anymore.
Was this copied from previous nix stuff as i see nix related files above as well?

File: .github/workflows/release-please.yml [hunk 5]
```diff
@@ -26,3 +29,100 @@ jobs:
...
+        run: |
+          cd cpp
+          cmake --preset default -DCMAKE_BUILD_TYPE=RelWithAssert -DOMP_MULTITHREADING=OFF
+          cmake --build --preset default ${@/#/--target }
+      - name: Compile Barretenberg.js
+        run: |
+          cd ts
+          yarn && yarn build
+      - name: Tar Typescript `dest` folder
+        run: tar -cvf dest.tar ts/dest
```
Comment (charlielye) [inline .github/workflows/release-please.yml]: These should be tgz's no?

File: .github/workflows/release-please.yml [hunk 6]
```diff
@@ -26,3 +29,100 @@ jobs:
...
+      - name: Tar Native static `lib` folder
+        run: tar -cvf ubuntu-static-lib.tar cpp/build/lib
```
Comment (charlielye) [inline .github/workflows/release-please.yml]: How come outputting static libs? Were we not going to output the artefects discussed in the thread?

File: .github/workflows/release-please.yml [hunk 7]
```diff
@@ -26,3 +29,100 @@ jobs:
...
+  build-mac:
+    name: Build on Mac
+    needs: [release-please]
+    runs-on: macos-latest
+    steps:
+      - name: Checkout
+        uses: actions/checkout@v3
+      - name: Create Mac Build Environment
+        run: |
+          brew install cmake llvm ninja libomp clang-format
```
Comment (charlielye) [inline .github/workflows/release-please.yml]: omp not needed. we're meant to be using stock llvm now.

**Review by charlielye — DISMISSED** [review]
It seems there's a lot we need to change here so maybe easier to discuss in person.

File: .github/workflows/release-please.yml (hunk 7 above)
Comment (kevaundray) [inline .github/workflows/release-please.yml]:
```suggestion
          brew install cmake llvm ninja clang-format
```

File: .github/workflows/release-please.yml [hunk 8]
```diff
@@ -26,3 +29,100 @@ jobs:
...
+              sudo apt -y remove --purge cmake
+              sudo snap install cmake --classic
+      - name: Install Yarn
+        run: |
+          curl -sS https://dl.yarnpkg.com/debian/pubkey.gpg | sudo apt-key add -
+          echo "deb https://dl.yarnpkg.com/debian/ stable main" | sudo tee /etc/apt/sources.list.d/yarn.list
+          sudo apt -y update && sudo apt -y install yarn
+      - name: Install Clang/Ninja/Binaryen
+        run: |
+          sudo apt-get -y install clang clang-format ninja-build binaryen
```
Comment (kevaundray) [inline .github/workflows/release-please.yml]:
```suggestion
          sudo apt-get -y install clang clang-format ninja-build
```

File: .github/workflows/release-please.yml (hunk 4 above)
Comment (kevaundray) [inline .github/workflows/release-please.yml]:
```suggestion
      - name: Install Clang/Ninja
```

File: .github/workflows/release-please.yml [hunk 9]
```diff
@@ -26,3 +29,100 @@ jobs:
...
+          sudo apt-get -y install clang clang-format ninja-build
+      - name: Install WASI-SDK
+        run: |
+          cd cpp
+          ./scripts/install-wasi-sdk.sh
+      - name: Compile Barretenberg
+        run: |
+          cd cpp
+          cmake --preset default -DCMAKE_BUILD_TYPE=RelWithAssert -DOMP_MULTITHREADING=OFF
+          cmake --build --preset default ${@/#/--target }
```
Comment (kevaundray) [inline .github/workflows/release-please.yml]:
```suggestion
          cmake --build --preset default --target bb
```

File: .github/workflows/release-please.yml [hunk 10]
```diff
@@ -26,3 +29,100 @@ jobs:
...
+      - name: Create Mac Build Environment
+        run: |
+          brew install cmake llvm ninja clang-format
+      - name: Configure Brew prefix
+        run: echo "BREW_PREFIX=$(brew --prefix)" >> $GITHUB_ENV
+      - name: Compile Barretenberg
+        run: |
+          cd cpp
+          cmake --preset default -DCMAKE_BUILD_TYPE=RelWithAssert -DOMP_MULTITHREADING=OFF
+          cmake --build --preset default ${@/#/--target }
```
Comment (kevaundray) [inline .github/workflows/release-please.yml]:
```suggestion
          cmake --build --preset default -- target bb
```

File: .github/workflows/release-please.yml [hunk 11]
```diff
@@ -26,3 +29,100 @@ jobs:
...
+        uses: actions/checkout@v3
+      - name: Create Mac Build Environment
+        run: |
+          brew install cmake llvm ninja clang-format
+      - name: Configure Brew prefix
+        run: echo "BREW_PREFIX=$(brew --prefix)" >> $GITHUB_ENV
+      - name: Compile Barretenberg
+        run: |
+          cd cpp
+          cmake --preset default -DCMAKE_BUILD_TYPE=RelWithAssert -DOMP_MULTITHREADING=OFF
```
Comment (kevaundray) [inline .github/workflows/release-please.yml]:
```suggestion
          cmake --preset default -DCMAKE_BUILD_TYPE=RelWithAssert
```

File: .github/workflows/release-please.yml [hunk 12]
```diff
@@ -26,3 +29,100 @@ jobs:
...
+  build-mac:
+    name: Build on Mac
+    needs: [release-please]
+    runs-on: macos-latest
+    steps:
+      - name: Checkout
+        uses: actions/checkout@v3
+      - name: Create Mac Build Environment
+        run: |
+          brew install cmake llvm ninja clang-format
```
Comment (kevaundray) [inline .github/workflows/release-please.yml]:
```suggestion
          brew install cmake ninja
```

**Comment by kevaundray** [comment]
token name: `NPM_TOKEN`

File: .github/workflows/release-please.yml [hunk 13]
```diff
@@ -26,3 +31,124 @@ jobs:
...
+    - name: Upload artifacts
+      uses: actions/upload-artifact@v2
+      with:
+          name: release-linux-wasm
+          path: |
+            barretenberg.wasm
+            bb-ubuntu.tar.gz
+          
+  build-mac:
+    name: Build on Mac
```
Comment (kevaundray) [inline .github/workflows/release-please.yml:93]: Blocking: Mac builds are currently not compiling

File: .github/workflows/release-please.yml [hunk 14]
```diff
@@ -26,3 +31,130 @@ jobs:
...
+      - name: Select Xcode version
+        run: |
+         sudo xcode-select -switch /Applications/Xcode_14.3.1.app
+
+      - name: Create Mac Build Environment
+        run: |
+          brew install cmake ninja
+      - name: Configure Brew prefix
+        run: |
+          echo "BREW_PREFIX=$(brew --prefix)" >> $GITHUB_ENV
```
Comment (kevaundray) [inline .github/workflows/release-please.yml]:
```suggestion
```

File: .github/workflows/release-please.yml (hunk 13 above)
Comment (kevaundray) [reply .github/workflows/release-please.yml:93]: This is now fixed by using macos-13 ventura and upgrading clang to 14.0.3 (xcode 14.3.1)

**Pull request merged**